backend/controller/controller.go

```diff
@@ -1390,78 +1409,127 @@ func (s *Service) executeAsyncCalls(ctx context.Context) (time.Duration, error)
 
 	logger = logger.Scope(fmt.Sprintf("%s:%s", call.Origin, call.Verb))
 
+	if call.Catching {
+		// Retries have been exhausted but catch verb has previously failed
+		// We need to try again to catch the async call
+		return 0, s.catchAsyncCall(ctx, logger, call, queueDepth)
+	}
+
 	logger.Tracef("Executing async call")
 	req := &ftlv1.CallRequest{
 		Verb: call.Verb.ToProto(),
 		Body: call.Request,
 	}
 	resp, err := s.callWithRequest(ctx, connect.NewRequest(req), optional.None[model.RequestKey](), s.config.Advertise.String())
 	var callResult either.Either[[]byte, string]
-	failed := false
 	if err != nil {
 		logger.Warnf("Async call could not be called: %v", err)
-		observability.AsyncCalls.Executed(ctx, call.Verb, call.Origin.String(), call.ScheduledAt, optional.Some("async call could not be called"))
+		observability.AsyncCalls.Executed(ctx, call.Verb, call.CatchVerb, call.Origin.String(), call.ScheduledAt, false, optional.Some("async call could not be called"))
 		callResult = either.RightOf[[]byte](err.Error())
-		failed = true
 	} else if perr := resp.Msg.GetError(); perr != nil {
 		logger.Warnf("Async call failed: %s", perr.Message)
-		observability.AsyncCalls.Executed(ctx, call.Verb, call.Origin.String(), call.ScheduledAt, optional.Some("async call failed"))
+		observability.AsyncCalls.Executed(ctx, call.Verb, call.CatchVerb, call.Origin.String(), call.ScheduledAt, false, optional.Some("async call failed"))
 		callResult = either.RightOf[[]byte](perr.Message)
-		failed = true
 	} else {
 		logger.Debugf("Async call succeeded")
 		callResult = either.LeftOf[string](resp.Msg.GetBody())
-		observability.AsyncCalls.Executed(ctx, call.Verb, call.Origin.String(), call.ScheduledAt, optional.None[string]())
+		observability.AsyncCalls.Executed(ctx, call.Verb, call.CatchVerb, call.Origin.String(), call.ScheduledAt, false, optional.None[string]())
 	}
-	err = s.dal.CompleteAsyncCall(ctx, call, callResult, func(tx *dal.Tx) error {
-		if failed && call.RemainingAttempts > 0 {
-			// If the call failed and there are attempts remaining, then
-			// CompleteAsyncCall would enqueue another call, so increment
-			// queue depth accordingly.
-			queueDepth++
 
-			// Will retry, do not propagate failure yet.
-			return nil
-		}
-		// Allow for handling of completion based on origin
-		switch origin := call.Origin.(type) {
-		case dal.AsyncOriginFSM:
-			return s.onAsyncFSMCallCompletion(ctx, tx, origin, failed)
+	didScheduleAnotherCall, err := s.dal.CompleteAsyncCall(ctx, call, callResult, func(tx *dal.Tx, isFinalResult bool) error {
+		return s.finaliseAsyncCall(ctx, tx, call, callResult, isFinalResult)
+	})
+	if err != nil {
+		observability.AsyncCalls.Completed(ctx, call.Verb, call.CatchVerb, call.Origin.String(), call.ScheduledAt, false, queueDepth, err)
+		return 0, fmt.Errorf("failed to complete async call: %w", err)
+	}
+	if didScheduleAnotherCall {
+		// Increment queue depth to model the next scheduled attempt
+		queueDepth++
+	}
```

Comment:
Should we move this check up to above the `err != nil` check? Just since the presence of `err` no longer relates to whether another call was scheduled; i.e. `didScheduleAnotherCall` will always be correct regardless of whether `err == nil`.

Technically, it makes no difference in this PR, because for every case where we return an error, `didScheduleAnotherCall` would simultaneously be false, but I'm wondering if that would be more durable in the long run because _if_ anyone added a case to `CompleteAsyncCall` that returned `true, err`, then `observability.AsyncCalls.Completed` would receive the wrong `queueDepth`.

On the other hand, performing the `err != nil` check first is more idiomatic, and the chances of us introducing a return case where `didScheduleAnotherCall == true` and `err != nil` is unlikely. I don't have strong feelings either way - totally up to you.

Reply:
I tried something a bit different because now that we changed other things, i don't think we need to decrement the counter immediately. So instead of doing `--` and then `++` in some cases, instead we just decrement when the call has ended. That means the error case should be safer because neither the call depth or the async call table has been updated
[only decrement queue depth once we know we ended the async call](https://github.com/TBD54566975/ftl/pull/2190/commits/41f9f6d77760a927b0c70680ed7cb4e20f5f7ed6)